dataverse-webapp/src/test/java/edu/harvard/iq/dataverse/dataverse/themewidget/ThemeWidgetServiceIT.java

```diff
@@ -0,0 +1,123 @@
+    @EJB
+    private AuthenticationServiceBean authenticationServiceBean;
+
+    @Before
+    public void setUp() {
+        dataverseSession.setUser(authenticationServiceBean.getAdminUser());
+    }
+
+
+    @Test
+    public void shouldSaveOrUpdateThemeRoot() {
+        // given
+        Dataverse dataverse = dataverseDao.findByAlias("ownmetadatablocks");
+        File file = new File (getClass().getClassLoader()
+                .getResource("images/banner.png").getFile());
+
+        DataverseTheme dvTheme = dataverse.getDataverseTheme();
+        dvTheme.setLogo("banner.png");
+        dvTheme.setLogoFormat(DataverseTheme.ImageFormat.RECTANGLE);
+        dvTheme.setLinkUrl("http://icm.edu.pl");
+        dvTheme.setLogoAlignment(DataverseTheme.Alignment.RIGHT);
+        dvTheme.setBackgroundColor("FFFFAA");
+        dvTheme.setLogoBackgroundColor("FFFFAA");
+        dvTheme.setLinkColor("FFFFAA");
+        dvTheme.setTextColor("FFFFAA");
+        dvTheme.setTagline("testTagline");
+
+        // when
+        themeWidgetService.saveOrUpdateThemeRoot(dataverse, file);
+
+        // then
+        Dataverse dbDataverse = dataverseDao.findByAlias("ownmetadatablocks");
+        DataverseTheme dbDvTheme = dbDataverse.getDataverseTheme();
+
+        assertEquals("Root", dbDataverse.getThemeRootDataverseName());
+        assertEquals("banner.png", dbDvTheme.getLogo());
+        assertEquals(DataverseTheme.ImageFormat.RECTANGLE, dbDvTheme.getLogoFormat());
+        assertEquals(DataverseTheme.Alignment.RIGHT, dbDvTheme.getLogoAlignment());
+        assertEquals("http://icm.edu.pl", dbDvTheme.getLinkUrl());
+        assertEquals("FFFFAA", dbDvTheme.getBackgroundColor());
+        assertEquals("FFFFAA", dbDvTheme.getLogoBackgroundColor());
+        assertEquals("FFFFAA", dbDvTheme.getLinkColor());
+        assertEquals("FFFFAA", dbDvTheme.getTextColor());
+        assertEquals("testTagline", dbDvTheme.getTagline());
+
+        Path logoPath = Paths.get("../docroot/logos");
+        File logoFileDir = new File(logoPath.toFile(), dbDataverse.getId().toString());
+        File newFile = new File(logoFileDir, dbDataverse.getDataverseTheme().getLogo());
+
+        assertTrue(logoFileDir.exists());
+        assertTrue(logoFileDir.isDirectory());
+        assertTrue(newFile.exists());
+        assertTrue(newFile.isFile());
+        assertEquals("banner.png", newFile.getName());
+    }
+
+    @Test
+    public void shouldSaveOrUpdateTheme() {
+        // given
+        Dataverse dataverse = dataverseDao.findByAlias("unreleased");
+        dataverse.setThemeRoot(false);
+        em.persist(dataverse);
+
+
+        // when
+        themeWidgetService.saveOrUpdateTheme(dataverse);
+
+        // then
+        Dataverse dbDataverse = dataverseDao.findByAlias("unreleased");
+        assertEquals("Root", dbDataverse.getThemeRootDataverseName());
+        assertEquals(dataverseDao.findByAlias("root").getDataverseTheme(), dbDataverse.getDataverseTheme());
+
```

Comment:
Shouldn't we check if dataverse have/doesn't have theme since we updated it?

Reply:
I guess method naming was confusing. The Page class requires a proper refactor since both design is questionable and logic is VERY fuzzy it's easy getting lost. 
What really happens in this place is: we inherit theme from root dataverse. 
What we can make assertion against is what is already being done - check if the dataverse has the same theme as its root (which is Root dataverse).